pkg/destroy/powervs/powervs.go

```diff
@@ -0,0 +1,956 @@
+		var wg sync.WaitGroup
+		errCh := make(chan error)
+		wgDone := make(chan bool)
+
+		for _, f := range stage {
+			wg.Add(1)
+			// Start a parallel goroutine
+			go o.executeStageFunction(f, errCh, &wg)
+		}
+
+		// Start a parallel goroutine
+		go func() {
+			wg.Wait()
+			close(wgDone)
+		}()
+
+		select {
+		// Did the wait group goroutine finish?
+		case <-wgDone:
+			// On to the next stage
+			o.Logger.Debugf("destroyCluster: <-wgDone")
+			continue
+		// Have we taken too long?
+		case <-time.After(stageTimeout):
+			return fmt.Errorf("destroyCluster: timed out")
+		// Has an error been sent via the channel?
+		case err := <-errCh:
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *ClusterUninstaller) executeStageFunction(f struct {
+	name    string
+	execute func() error
+}, errCh chan error, wg *sync.WaitGroup) error {
+	o.Logger.Debugf("executeStageFunction: Adding: %s", f.name)
+
+	defer wg.Done()
+
+	var (
+		ctx      context.Context
+		deadline time.Time
+		ok       bool
+		err      error
+	)
+
+	ctx, _ = o.contextWithTimeout()
+	if ctx == nil {
+		return errors.Wrap(err, "executeStageFunction contextWithTimeout returns nil")
+	}
+
+	deadline, ok = ctx.Deadline()
+	if !ok {
+		return errors.Wrap(err, "executeStageFunction failed to call ctx.Deadline")
+	}
+
+	var duration time.Duration = deadline.Sub(time.Now())
+
+	o.Logger.Debugf("executeStageFunction: duration = %v", duration)
+	if duration <= 0 {
+		return fmt.Errorf("executeStageFunction: duration is <= 0 (%v)", duration)
+	}
+
+	err = wait.PollImmediateInfinite(
+		duration,
+		func() (bool, error) {
+			var err error
+
+			o.Logger.Debugf("executeStageFunction: Executing: %s", f.name)
+
+			err = f.execute()
+			if err != nil {
+				o.Logger.Debugf("ERROR: executeStageFunction: %s: %v", f.name, err)
+
+				return false, err
+			}
+
+			return true, nil
+		},
+	)
+
+	if err != nil {
+		errCh <- err
+	}
+	return nil
+}
+
+// GetCISInstanceCRN gets the CRN name for the specified base domain.
+func GetCISInstanceCRN(BaseDomain string) (string, error) {
+	var (
+		CISInstanceCRN                string = ""
+		APIKey                        string
+		bxSession                     *bxsession.Session
+		err                           error
+		tokenProviderEndpoint         string = "https://iam.cloud.ibm.com"
+		tokenRefresher                *authentication.IAMAuthRepository
+		authenticator                 *core.IamAuthenticator
+		controllerSvc                 *resourcecontrollerv2.ResourceControllerV2
+		listInstanceOptions           *resourcecontrollerv2.ListResourceInstancesOptions
+		listResourceInstancesResponse *resourcecontrollerv2.ResourceInstancesList
+		instance                      resourcecontrollerv2.ResourceInstance
+		zonesService                  *zonesv1.ZonesV1
+		listZonesOptions              *zonesv1.ListZonesOptions
+		listZonesResponse             *zonesv1.ListZonesResp
+	)
+
+	if APIKey = os.Getenv("IBMCLOUD_API_KEY"); len(APIKey) == 0 {
```

Comment:
Same comment as above, this should be stored in a session not an env var.

Reply:
i believe these have been taken care of in a recent push (though haven't looked). @hamzy can you confirm?

Reply:
Yes, it has been fixed.